### DISCUSSION TITLE
Distil-Whisper Models: 6 times speed up and 49% smaller size?

### DISCUSSION THREAD

**Discussion by lukestanley** [ggml-org/whisper.cpp#1414] [Show and tell]
Tomorrow, HuggingFace's team are set to release their distilled Whisper models, which claim to be "6 times faster, 49% smaller, and perform within 1% WER on out-of-distribution evaluation sets." You should be able to find them here **tomorrow**: [HuggingFace Collection](https://huggingface.co/collections/distil-whisper/distil-whisper-models-65411987e6727569748d2eb6) and [GitHub](https://github.com/huggingface/distil-whisper).
If they're in PT format, I assume that this script should do the trick, perhaps better minds can confirm tomorrow: [Convert PT to GGML Script](https://github.com/ggerganov/whisper.cpp/blob/master/models/convert-pt-to-ggml.py). This should bring about even faster speed improvements. For those of us (like me) with limited hardware, we might get high-quality ASR on our own devices that's faster than real-time, and doesn't rely on network connectivity. So, tomorrow should be a good day!

**Comment by ggerganov** [comment]
Exciting! 

Edit: initial support has been added in this PR - https://github.com/ggerganov/whisper.cpp/pull/1424

Please note that the current implementation is not optimal since it does not support the proposed chunking strategy (see https://github.com/ggerganov/whisper.cpp/discussions/1414#discussioncomment-7465861 for more info). Therefore, the transcription quality using `whisper.cpp` compared to the original `distil-whisper` transformers implementation can be lower in some cases.

**Reply by bobqianic** [reply]
It looks like the initial release is limited to English since the training set consisted solely of English audio. Check this out for further information: https://github.com/huggingface/distil-whisper/issues/6#issuecomment-1790869897

**Reply by bobqianic** [reply]
In README the author mentioned: 

> Distil-Whisper is a distilled version of Whisper for English speech recognition that is 6 times faster, 49% smaller, and performs within 1% word error rate (WER) on out-of-distribution evaluation sets. Multilingual support will be provided soon through distillation training code.

**Comment by jordibruin** [comment]
They should be released now. Curious to see performance metrics 🙂

**Comment by AdithyanI** [comment]
They are released : https://www.linkedin.com/posts/sanchit-gandhi_distil-whisper-is-now-available-in-transformers-activity-7125902709794189313-aumY?utm_source=share&utm_medium=member_desktop


Anyone already ran the model? Would love to see some results :) 
I am not deep into this yet, otherwise would do it myself.

**Reply by fblissjr** [reply]
Tested it with speculative decoding per HF's example on the model page (https://huggingface.co/distil-whisper/distil-large-v2). Pretty happy with results.

**Reply by bobqianic** [reply]
I've tested with whisper.cpp, and it seems that the automatic language detection isn't working—the distilled model can't correctly identify the language used in the audio.

**Reply by patrickvonplaten** [reply]
The models are all English-only - we should probably clarify this better in the model card.
~co-author

**Comment by AdithyanI** [comment]
How do I convert this to something whisper.cpp can use? 
I tried following this : https://github.com/ggerganov/whisper.cpp/tree/master/models

But I cannot find any .pt files in here : https://huggingface.co/distil-whisper/distil-medium.en/tree/main

**Reply by lukestanley** [reply]
I'm not sure but I suspect that this converter might be more appropriate:
https://github.com/ggerganov/whisper.cpp/blob/master/models/convert-h5-to-ggml.py
@AdithyanI

**Reply by bobqianic** [reply]
No, the distilled model is actually provided in an h5 format, which means there isn't any `.pt` files.

**1.** Start by downloading the full OpenAI Whisper repository. This is important as it includes essential files, such as the `mel_filter`, that you’ll need later on.
**2.** Next, go ahead and download everything from the following Hugging Face repository: https://huggingface.co/distil-whisper/distil-large-v2/tree/main. Make sure you get all the contents.
**3.** Once you have those, navigate to the `model` directory in **whisper.cpp** and use the `convert-h5-to-ggml` tool provided there to transform the `.h5` model files into `.ggml` format.

**Reply by AdithyanI** [reply]
Thanks, checking that now...

**Reply by AdithyanI** [reply]
<img width="1233" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/12037164/293ed47d-ec19-4131-80c8-d2d5e7e978a8">


Thanks @bobqianic . Total noob here, so everything is new.  

I followed the instructions in there : https://github.com/ggerganov/whisper.cpp/blob/master/models/convert-h5-to-ggml.py
And I got a .bin out of it. And I used that now instead of the regular base.en. It seems to run. But no output is generated. i.e. I don't see the text. 

I will continue debugging.


I used this : 

- https://huggingface.co/distil-whisper/distil-medium.en, and 
- https://github.com/openai/whisper

**Reply by AdithyanI** [reply]
This is what I see with "base.en"


<img width="1233" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/12037164/7f2bba29-7098-4ffc-aef9-159fef4dc1be">

**Reply by lukestanley** [reply]
Yeah I get the same error. To be clear, the standard "make tiny.en" worked fine when ran in the same way as the converted model.

Steps to reproduce with logs:

mkdir fast_asr
cd fast_asr/
git clone https://github.com/openai/whisper
git clone https://github.com/ggerganov/whisper.cpp
git clone https://huggingface.co/distil-whisper/distil-medium.en
python3 ./whisper.cpp/models/convert-h5-to-ggml.py ./distil-medium.en/ ./whisper .
cd whisper.cpp
make
make tiny.en
user@computer:~/fast_asr/whisper.cpp$ ./main -m /home/user/fast_asr/whisper.cpp/models/ggml-tiny.en.bin -f samples/jfk.wav -ml 16
whisper_init_from_file_no_state: loading model from '/home/user/fast_asr/whisper.cpp/models/ggml-tiny.en.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 384
whisper_model_load: n_audio_head  = 6
whisper_model_load: n_audio_layer = 4
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 384
whisper_model_load: n_text_head   = 6
whisper_model_load: n_text_layer  = 4
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 1
whisper_model_load: adding 1607 extra tokens
whisper_model_load: model ctx     =   73.62 MB
whisper_model_load: model size    =   73.54 MB
whisper_init_state: kv self size  =    2.62 MB
whisper_init_state: kv cross size =    8.79 MB
whisper_init_state: compute buffer (conv)   =   11.17 MB
whisper_init_state: compute buffer (encode) =   61.76 MB
whisper_init_state: compute buffer (cross)  =    3.67 MB
whisper_init_state: compute buffer (decode) =   18.82 MB

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | COREML = 0 | OPENVINO = 0 | 

main: processing 'samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...


[00:00:00.000 --> 00:00:00.940]   And so my
[00:00:00.940 --> 00:00:01.750]   fellow
[00:00:01.750 --> 00:00:04.000]   Americans ask
[00:00:04.000 --> 00:00:05.970]   not what your
[00:00:05.970 --> 00:00:06.880]   country can do
[00:00:06.880 --> 00:00:08.000]   for you
[00:00:08.000 --> 00:00:09.030]   ask what you
[00:00:09.030 --> 00:00:10.260]   can do for your
[00:00:10.260 --> 00:00:11.000]   country.


whisper_print_timings:     load time =   100.34 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    21.30 ms
whisper_print_timings:   sample time =    10.52 ms /    27 runs (    0.39 ms per run)
whisper_print_timings:   encode time =   683.69 ms /     1 runs (  683.69 ms per run)
whisper_print_timings:   decode time =    87.84 ms /    27 runs (    3.25 ms per run)
whisper_print_timings:   prompt time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:    total time =   947.55 ms
user@computer:~/fast_asr/whisper.cpp$ ./main -m /home/user/fast_asr/ggml-model.bin -f samples/jfk.wav -ml 16
whisper_init_from_file_no_state: loading model from '/home/user/fast_asr/ggml-model.bin'
whisper_model_load: loading model
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 1024
whisper_model_load: n_audio_head  = 16
whisper_model_load: n_audio_layer = 24
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 1024
whisper_model_load: n_text_head   = 16
whisper_model_load: n_text_layer  = 2
whisper_model_load: n_mels        = 80
whisper_model_load: ftype         = 1
whisper_model_load: qntvr         = 0
whisper_model_load: type          = 4
whisper_model_load: adding 1607 extra tokens
whisper_model_load: model ctx     =  756.86 MB
whisper_model_load: model size    =  756.66 MB
whisper_init_state: kv self size  =    3.50 MB
whisper_init_state: kv cross size =   11.72 MB
whisper_init_state: compute buffer (conv)   =   25.82 MB
whisper_init_state: compute buffer (encode) =  162.24 MB
whisper_init_state: compute buffer (cross)  =    7.33 MB
whisper_init_state: compute buffer (decode) =   47.74 MB

system_info: n_threads = 4 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | METAL = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | COREML = 0 | OPENVINO = 0 | 

main: processing 'samples/jfk.wav' (176000 samples, 11.0 sec), 4 threads, 1 processors, lang = en, task = transcribe, timestamps = 1 ...



whisper_print_timings:     load time =   572.91 ms
whisper_print_timings:     fallbacks =   0 p /   0 h
whisper_print_timings:      mel time =    18.31 ms
whisper_print_timings:   sample time =     0.87 ms /     2 runs (    0.44 ms per run)
whisper_print_timings:   encode time = 14216.68 ms /     1 runs (14216.68 ms per run)
whisper_print_timings:   decode time =    15.56 ms /     2 runs (    7.78 ms per run)
whisper_print_timings:   prompt time =     0.00 ms /     1 runs (    0.00 ms per run)
whisper_print_timings:    total time = 14897.49 ms

**Reply by lukestanley** [reply]
@bobqianic what did you do differently to make your working ggml model? I wonder if there is a hidden dependency issue, e.g with Transformers?

**Reply by bobqianic** [reply]
> what did you do differently to make your working ggml model?

I'm using the distilled `large-v2`, not the `medium.en`. I just did a quick test and it seems I'm not able to get any output from the `medium.en` either.

<img width="865" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/8a99c115-4481-4a9a-8406-5344fe2cae94">

**Reply by patrickvonplaten** [reply]
If it's still relevant we also added the checkpoints in original OAI format here:
- https://huggingface.co/distil-whisper/distil-large-v2/blob/main/original-large-32-2-en.bin
- https://huggingface.co/distil-whisper/distil-medium.en/blob/main/original-medium-24-2-en.bin

**Reply by patrickvonplaten** [reply]
.cpp checkpoints have also been added here:
[FP32]
- a) https://huggingface.co/distil-whisper/distil-large-v2/blob/main/ggml-large-32-2.fp32.en.bin
- b) https://huggingface.co/distil-whisper/distil-medium.en/blob/main/ggml-medium-32-2.en.fp32.bin
[FP16]
- a) https://huggingface.co/distil-whisper/distil-large-v2/blob/main/ggml-large-32-2.en.bin
- b) https://huggingface.co/distil-whisper/distil-medium.en/blob/main/ggml-medium-32-2.en.bin

**Comment by bobqianic** [comment]
I ran a few initial tests using `whisper.cpp` and it seems that even when using the `distil-large-v2` model, **_there's an issue with correctly transcribing audio that's not in English_**. The output from the model is garbage. I'm not entirely sure whether this is an issue with the `whisper.cpp` code or with the model itself. @ggerganov 

**Japanese:**
<img width="867" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/71c3cc08-e282-4b7f-9f15-8bc253616881">

**Chinese:**
<img width="866" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/2469338a-eb8f-465c-9a21-f58d906a32f3">

**English:**
<img width="867" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/129547291/03813aa0-43ac-4d85-9e7e-0e7c7e578217">

**Reply by AdithyanI** [reply]
Could you please share the .bin file with me? 
For the medium.en? Would hugely appreciate it.

**Reply by bobqianic** [reply]
https://huggingface.co/bobqianic/whisper.cpp-distilled/tree/main

**Reply by ficapy** [reply]
I tried using `ggml-model-distilled-medium.en.bin`, and the ASR speed increased by threefold. However, it produced incorrect results and omitted quite a lot of content.
On the M1, for 33 minutes of English language content, the original version took **297 seconds**, while the distilled version only took **100 seconds**.
```
yt-dlp -o "example.%(ext)s" --format "bestaudio" "https://www.youtube.com/watch?v=LQoQ7bEjYxg"
ffmpeg -i example.webm -threads 16 -acodec pcm_s16le -ac 1 -ar 16000 -vn example.wav -y
wget "https://huggingface.co/bobqianic/whisper.cpp-distilled/resolve/main/ggml-model-distilled-medium.en.bin"
./main -m /tmp/ggml-model-distilled-medium.en.bin -f example.wav -osrt
```
<img width="2261" alt="image" src="https://github.com/ggerganov/whisper.cpp/assets/3191500/e7cad4c3-1023-4e0e-8b45-935241eb8602">

**Reply by ggerganov** [reply]
Can you show the results with `-ps` added? ~Can you run the same audio with the reference distilled implementation in transformers and compare the results?~

**Reply by AdithyanI** [reply]
@ficapy thx. I confirm that too. I see a lot of missing content with medium.en.

**Reply by ggerganov** [reply]
I'm trying `medium.en-distill` and it looks like it requires to insert a language id token and task token, even though it is not a multi-lingual model:

```diff
diff --git a/whisper.cpp b/whisper.cpp
index ccac6aa..44d46b3 100644
--- a/whisper.cpp
+++ b/whisper.cpp
@@ -4560,7 +4560,7 @@ int whisper_full_with_state(
 
     // these tokens determine the task that will be performed
     std::vector<whisper_token> prompt_init = { whisper_token_sot(ctx) };
-    if (whisper_is_multilingual(ctx)) {
+    if (true) {
         const int lang_id = whisper_lang_id(params.language);
         state->lang_id = lang_id;
         prompt_init.push_back(whisper_token_lang(ctx, lang_id));
```

This at least starts producing some semi-good results, but still the quality is heavily lacking, so there might be some additional adjustments needed:

```bash
make -j && ./main -m models/ggml-medium.en.bin -f samples/gb0.wav -ps
```

![image](https://github.com/ggerganov/whisper.cpp/assets/1991296/6b557edb-8e0a-4d82-9cf1-6588694def1e)

The decoder speed with the distilled medium model is `~2.0 ms/t` which is more than x6 times faster than the OG model at `~13.2 ms/t`. The encoder is the same speed.

So we can expect total processing speed-ups for greedy decoding anywhere between no speed-up (audio without speech, i.e. processing dominated by the encoder) up to x3 faster (audio with lot's of speech, in which the decoder takes significantly more time than the encoder). A faster decoder also means much more efficient beam search performance.

But in any case, we first have to figure out what is causing the degraded transcription quality.

P.S. For reference, this is the same example using the OG medium.en model:

![image](https://github.com/ggerganov/whisper.cpp/assets/1991296/25aada74-3a7d-449b-939c-03a95afb7397)

**Reply by ficapy** [reply]
the results from the official transformer are correct.
```
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "distil-whisper/distil-medium.en"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=False, use_safetensors=True
)
model.to(device)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    chunk_length_s=15,
    batch_size=16,
    torch_dtype=torch_dtype,
    device=device,
)

result = pipe("example.wav")
print(result['text'])
```

**Reply by sanchit-gandhi** [reply]
Hey @lukestanley and @ggerganov - super cool to see the effort to convert Distil-Whisper to Whisper cpp. To clarify your most recent question - the `distil-medium.en` model uses the same task tokens as `medium.en` (English-only):
```
'<|startoftranscript|><|notimestamps|> Mr Quilter is the Apostle of the Middle Classes and we are glad to welcome his Gospel.<|endoftext|>'
```

Whereas `distil-large-v2` is inherited from the `large-v2` (multilingual) model, and so also inserts the task and language tokens:
```
'<|startoftranscript|><|en|><|transcribe|><|notimestamps|> Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel.<|endoftext|>'
```

**Reply by ggerganov** [reply]
So I am reading the paper and I realize 2 things:

- these models do not have timestamp prediction capability
- they require new custom chunk-based transcription strategy

![image](https://github.com/ggerganov/whisper.cpp/assets/1991296/79726c56-a54c-4646-9275-57221314ae11)

![image](https://github.com/ggerganov/whisper.cpp/assets/1991296/aa8922e7-b895-413c-a35a-35b076cafbf3)

Adding the `no_timestamp` to the initial sequence fixes both `medium` and `large` distill, though the quality at the chunk boundaries is obviously degraded:

```diff
diff --git a/whisper.cpp b/whisper.cpp
index ccac6aa..9e00513 100644
--- a/whisper.cpp
+++ b/whisper.cpp
@@ -3942,6 +3942,7 @@ static void whisper_process_logits(
         // suppress task tokens
         logits[vocab.token_translate]  = -INFINITY;
         logits[vocab.token_transcribe] = -INFINITY;
+        logits[vocab.token_prev]       = -INFINITY;
 
         if (params.logits_filter_callback) {
             params.logits_filter_callback(&ctx, &state, tokens_cur.data(), tokens_cur.size(), logits.data(), params.logits_filter_callback_user_data);
@@ -4560,7 +4561,8 @@ int whisper_full_with_state(
 
     // these tokens determine the task that will be performed
     std::vector<whisper_token> prompt_init = { whisper_token_sot(ctx) };
-    if (whisper_is_multilingual(ctx)) {
+    //if (whisper_is_multilingual(ctx)) { // large-distill
+    if (true) {                           // medium.en-distill
         const int lang_id = whisper_lang_id(params.language);
         state->lang_id = lang_id;
         prompt_init.push_back(whisper_token_lang(ctx, lang_id));
@@ -4569,6 +4571,7 @@ int whisper_full_with_state(
         } else {
             prompt_init.push_back(whisper_token_transcribe(ctx));
         }
+        prompt_init.push_back(whisper_token_not(ctx));
     }
 
     int seek = seek_start;
```

```bash
make -j && ./main -m models/ggml-medium.en-distill.bin -f samples/gb0.wav -nt -pc -ps
```

![image](https://github.com/ggerganov/whisper.cpp/assets/1991296/a46f2d96-928c-4867-af34-4d6956218599)

@ficapy Since you already have the python stuff setup, could you please run the medium distill version on the `samples/gb0.wav` sample audio in this repo (you might have to run `make samples` to download it, if you don't have it) and show me the transcription output that is produced?

**Reply by ggerganov** [reply]
@sanchit-gandhi Thanks, I just reverse-engineered that the missing `no_timestamps` token is necessary. However, `whisper.cpp` is currently lacking the chunking strategy with stride, so I'm not sure how to work around this. There will be degraded quality with the existing implementation

**Reply by ficapy** [reply]
@ggerganov  The results using transformer are as follows.


 

> Good morning. This Tuesday is election day. After months of spirited debate and vigorous campaigning, the time has come for Americans to make important decisions about our nation's future. I encourage all Americans to go to the polls and vote. Election season brings out the spirit of competition between our political parties. And that competition is an essential part of a healthy democracy. But as the campaigns come to a close, Republicans, Democrats, and independents can find common ground on at least one point. Our system of representative democracy is one of America's greatest strengths. The United States was founded on the belief that all men are created equal. Every election day, millions of Americans of all races, religions, and backgrounds step into voting booths throughout the nation. Whether they are rich or poor, old or young, each of them has an equal share in choosing the path that our country will take. And every ballot they cast is a reminder that our founding principles are alive and well. Voting is one of the great privileges of American citizenship, and it has always required brave defenders. As you head to the polls next week, remember the sacrifices that have been made by generations of Americans in uniform to preserve our way of life. From Bunker Hill to Baghdad, the men and women of American Armed Forces have been devoted guardians of our democracy. All of us owe them and their families a special debt of gratitude on election day. Americans should also remember the important example that our elections set throughout the world. Young democracies from Georgia and Ukraine to Afghanistan and Iraq can look to the United States for proof that self-government can endure, and nations that still live under tyranny and oppression can find hope and inspiration in our commitment to liberty. For more than two centuries Americans have demonstrated the ability of free people to choose their own leaders. Our nation has flourished because of its commitment to trusting the wisdom of our citizenry. In this year's election, we will see this tradition continue. And we will be reminded once again that we are blessed to live in a free nation guided by the will of the people. Thank you for listening.

 Additionally, according to the author's reply, it seems that timestamps can be generated. https://github.com/huggingface/distil-whisper/issues/13#issuecomment-1792139116

**Reply by patrickvonplaten** [reply]
Timestamps can also be generated with distil-whisper! In Transformers all you need to do is to set `return_token_timestaps=True` for generate: https://huggingface.co/docs/transformers/model_doc/whisper#transformers.WhisperForConditionalGeneration.generate.return_token_timestamps

**Reply by patrickvonplaten** [reply]
In our experiments adding timestamps however didn't improve transcription quality for distil-whisper (even for long-form evaluations). 

So we recommend adding timestamps only if you actually need to know the time, not if you're trying to improve quality this way.

**Comment by patrickvonplaten** [comment]
Amazing that you're already on it! Linking the other issue here: https://github.com/ggerganov/whisper.cpp/issues/1423

**Comment by gtubolcev** [comment]
Any news about distil-whisper models converstion?
